### PR TITLE
Remove `Dimension::Undefined` in favour of `Option<Dimension>`

### LINKED ISSUE #114: Consider removing Dimension::Undefined
My instincts suggest this should be an `Option<Dimension>`.

### LINKED ISSUE #154: Change `width` and `height` of `Size` to `Option<T>` instead of `T`
## What problem does this solve or what need does it fill?

The `geometry::Size<T>` is currently defined as
```rs
pub struct Size<T> {
    /// The x extent of the rectangle
    pub width: T,
    /// The y extent of the rectangle
    pub height: T,
}
```

While digging through the code in relation to #144 and #148 I've started wondering if we should change the definition of `Size` to the following:

```rs
pub struct Size<T> {
    /// The x extent of the rectangle
    pub width: Option<T>,
    /// The y extent of the rectangle
    pub height: Option<T>,
}
```

The width and height fields need to be independently defined or not, which means that replacing everything with `some_field: Option<Size>` will not cover all use-cases throughout the code. 

I'm not yet able to see the full picture how everything fits together once we:
`Dimension::Undefined` -> `Option<Dimension>` 
`Number::undefined` -> `Option<f32>`

but it seems it is a must for this to work (?)

## What alternative(s) have you considered?

Leave it as is, but I expect it to change in some way or another. 

Are there alternative designs we can explore?

### PR DESCRIPTION
**Pull request by Weibye**
# Objective

Fixes #114 
Fixes #154 

Enabled by the work from #144 

## Changelog

- `Dimension::Undefined` has been removed
- Which means we need to deal with undefined / defined nature of the dimensions in one level higher, using `Option<Dimension>`

### `Size`

The `geometry::Size<T>` has been changed from
```rs
pub struct Size<T> {
    /// The x extent of the rectangle
    pub width: T,
    /// The y extent of the rectangle
    pub height: T,
}
```
to
```rs
pub struct Size<T> {
    /// The x extent of the rectangle
    pub width: Option<T>,
    /// The y extent of the rectangle
    pub height: Option<T>,
}
```

### `Rect`

Has changed from 
```rs
pub struct Rect<T> {
    pub start: T,
    pub end: T,
    pub top: T,
    pub bottom: T,
}
```
to
```rs
pub struct Rect<T> {
    pub start: Option<T>,
    pub end: Option<T>,
    pub top: Option<T>,
    pub bottom: Option<T>,
}
```

## Feedback wanted

- TODO

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
@Weibye I've done my best to resolve the merge conflicts and rebase for you :)

File: src/geometry.rs [hunk 1]
```diff
@@ -67,7 +67,7 @@ where
     /// This is typically used when computing total padding.
     ///
     /// **NOTE:** this is *not* the width of the rectangle.
-    pub(crate) fn horizontal_axis_sum(&self) -> T {
+    pub(crate) fn horizontal_axis_sum(&self) -> Option<T> {
         self.start + self.end
```
Comment (Weibye) [inline src/geometry.rs:71]: How can I leverage the `MaybeMath` trait here?

Comment (alice-i-cecile) [reply src/geometry.rs]: Use .maybe_add

File: src/flexbox.rs [hunk 2]
```diff
@@ -95,7 +95,7 @@ struct AlgoConstants {
     padding_border: Rect<f32>,
 
     /// The size of the internal node
-    node_inner_size: Size<Option<f32>>,
+    node_inner_size: Size<f32>,
     /// The size of the surrounding container
     container_size: Size<f32>,
```
Comment (Weibye) [inline src/flexbox.rs:100]: Looking at this, do we need a `Size` that does not have Optional fields? One that is guaranteed to have both width and height?

File: src/flexbox.rs [hunk 3]
```diff
@@ -206,31 +203,29 @@ impl Forest {
...
 
-        let margin = self.nodes[node].style.margin.map(|n| n.resolve(parent_size.width).unwrap_or(0.0));
-        let padding = self.nodes[node].style.padding.map(|n| n.resolve(parent_size.width).unwrap_or(0.0));
-        let border = self.nodes[node].style.border.map(|n| n.resolve(parent_size.width).unwrap_or(0.0));
+        let margin =
+            self.nodes[node].style.margin.map(|n| n.map(|d| d.resolve(parent_size.width)).unwrap_or(Some(0.0)));
+        let padding =
+            self.nodes[node].style.padding.map(|n| n.map(|d| d.resolve(parent_size.width)).unwrap_or(Some(0.0)));
+        let border =
+            self.nodes[node].style.border.map(|n| n.map(|d| d.resolve(parent_size.width)).unwrap_or(Some(0.0)));
```
Comment (Weibye) [inline src/flexbox.rs:217]: I'm very uncertain if this is the correct logic. Still have more work to do before everything compiles and tests can start to yell at me.

File: src/flexbox.rs [hunk 4]
```diff
@@ -206,31 +203,29 @@ impl Forest {
...
+            end: padding.end.maybe_add(border.end),
+            top: padding.top.maybe_add(border.top),
+            bottom: padding.bottom.maybe_add(border.bottom),
         };
 
         let node_inner_size = Size {
-            width: node_size.width.maybe_sub(padding_border.horizontal_axis_sum()),
-            height: node_size.height.maybe_sub(padding_border.vertical_axis_sum()),
+            width: Some(node_size.width.maybe_sub(padding_border.horizontal_axis_sum())),
+            height: Some(node_size.height.maybe_sub(padding_border.vertical_axis_sum())),
```
Comment (Weibye) [inline src/flexbox.rs:228]: Compiler complains here: How can I satisfy the trait constraints of `MaybeMath` so that these sums work?

**Review by Weibye — COMMENTED** [review]
Making progress, uncertain of a few things

File: src/geometry.rs [hunk 5]
```diff
@@ -60,24 +62,26 @@ impl<T> Rect<T> {
 
 impl<T> Rect<T>
 where
-    T: Add<Output = T> + Copy + Clone,
+    T: MaybeMath<T, Option<T>> + Copy + Clone,
+    T: MaybeMath<Option<T>, Option<T>> + Copy + Clone,
+    T: MaybeMath<Option<T>, T> + Copy + Clone,
```
Comment (Weibye) [inline src/geometry.rs:67]: I did this to get the sums to work `MaybeMath` but it didn't solve it.

File: src/geometry.rs [hunk 6]
```diff
@@ -60,24 +62,26 @@ impl<T> Rect<T> {
...
 {
     /// The sum of [`Rect.start`](Rect) and [`Rect.end`](Rect)
     ///
     /// This is typically used when computing total padding.
     ///
     /// **NOTE:** this is *not* the width of the rectangle.
-    pub(crate) fn horizontal_axis_sum(&self) -> T {
-        self.start + self.end
+    pub(crate) fn horizontal_axis_sum(&self) -> Option<T> {
+        self.start.and_then(|start| start.maybe_add(self.end))
```
Comment (Weibye) [inline src/geometry.rs:75]: With `MaybeMath` I should be able to just use `self.start.maybe_add(self.end)` here (?) but not sure exactly how to get that to work. 

Traits this involved are new to me, so bear with me while I try to figure this out 🙃

**Pull request closed without merging**